jax/_src/lax/control_flow.py

```diff
@@ -1116,7 +1116,16 @@ def cond_bind(*args, branches, linear):
 
 ### scan
 
-def scan(f, init, xs, length=None, reverse=False, unroll=1):
+Carry = TypeVar('Carry')
+X = TypeVar('X')
+Y = TypeVar('Y')
+
+def scan(f: Callable[[Carry, X], Tuple[Carry, Y]],
+         init: Carry,
+         xs: Optional[Union[X, Sequence[X]]],
```

Comment:
Unfortunately this type is too restrictive. `scan` accepts arbitrary pytrees, not just values and sequences.

We also don't know how to describe pytrees to the Python type checkers, so I think the best we can do here is something like `Any` or something that is sugar for `Any`.

Reply:
Thanks for looking at this PR!  Although, I don't understand how `X` is too restrictive.  `X` is an unbound `TypeVar`, which means it is bound to `Any`.  Perhaps you could illustrate a valid call to scan that doesn't satisfy the annotation?

If anything, I wish `scan` were more restrictive accepting only:
```python
def scan(f: Callable[[Carry, X], Tuple[Carry, Y]],
         init: Carry,
         xs: Optional[X],
         length: Optional[int] = None,
         reverse: bool = False,
         unroll: int = 1) -> Tuple[Carry, Y]:
```
And for the cases where `xs` used to be a sequence, users could pass in `tree_multimap(jnp.stack, xs)` in that place.

Reply:
I'm think was confused by the `Sequence`. `X` can be pretty much anything (e.g., a dict of list of arrays is fine). But why isn't the type annotation just:

```
Carry = TypeVar('Carry')
X = TypeVar('X')
Y = TypeVar('Y')

def scan(f: Callable[[Carry, X], Tuple[Carry, Y]],
         init: Carry,
         xs: X,
         length: Optional[int] = None,
         reverse: bool = False,
         unroll: int = 1) -> Tuple[Carry, Y]:
```

You don't, as far as I can tell, need to special case either `Optional` or `Sequence`. `xs` can already be `None` or a sequence by virtue of being a free type variable.

Reply:
The reason that you need to special case `Optional` and `Sequence` is because the `TypeVar` `X` ensures that `xs` matches the argument to `f`.   With your definition, you won't be able to accept:
```python
scan(f, some_carry, None, 10)  # error: Cannot infer type argument 1 of "scan"
```
I'm perfectly happy removing the `Sequence` annotation if you prefer that.

Reply:
The sequence example you give first is not type correct, so there's nothing wrong with rejecting it. Arrays and lists are not interchangeable from the perspective of scan (or indeed any pytree-accepting function).

e.g.
```
import jax
from jax import lax

from typing import Tuple

def f(carry: float, x: int) -> Tuple[float, float]:
    print(carry)
    print(x)
    return carry, x + 0.5

lax.scan(f, 0.7, [1, 2, 3, 4])
```

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
~/p/jax/jax/_src/lax/control_flow.py in scan(f, init, xs, length, reverse, unroll)
   1196   try:
-> 1197     lengths = [x.shape[0] for x in xs_flat]
   1198   except AttributeError as err:

~/p/jax/jax/_src/lax/control_flow.py in <listcomp>(.0)
   1196   try:
-> 1197     lengths = [x.shape[0] for x in xs_flat]
   1198   except AttributeError as err:

AttributeError: 'int' object has no attribute 'shape'

The above exception was the direct cause of the following exception:

ValueError                                Traceback (most recent call last)
<ipython-input-3-0b5bbb8fa958> in <module>
      9     return carry, x + 0.5
     10
---> 11 lax.scan(f, 0.7, [1, 2, 3, 4])

~/p/jax/jax/_src/lax/control_flow.py in scan(f, init, xs, length, reverse, unroll)
   1200     raise ValueError(
   1201       msg.format(', '.join(str(x) for x in xs_flat
-> 1202                            if not hasattr(x, 'shape')))) from err
   1203
   1204   if length is not None:

ValueError: scan got value with no leading axis to scan over: 1, 2, 3, 4.
```

I'm also quite hesitant about the need to special case `Optional`. If anything I think that says "it might be better to just use `Any` instead of `X` and `Y`, because the type checker has no good way to understand pytrees".

Reply:
Oh, that's good to know.  I assumed that sequences would be handled by the flatten that happens inside `scan`.  I always use dataclasses containing arrays, so I never thought to try it.  Thanks for setting me straight!  I'll remove that now.

Regarding making `xs` optional, it seems logical to me.  Could you give me an example that should be accepted by the type variable approach, but isn't?  I think that would make it clear.

Reply:
Okay, you're right!  I misunderstood the way the flattening inside scan works.  Thanks for setting me straight.  I'll fix the annotation now.